File: 0408-valid-word-abbreviation/0408-valid-word-abbreviation.py
class Solution:
    def validWordAbbreviation(self, word: str, abbr: str) -> bool:
        word_ptr = 0
        abbr_ptr = 0
        
        while word_ptr < len(word) and abbr_ptr < len(abbr):
            if abbr[abbr_ptr].isdigit():
                if abbr[abbr_ptr] == "0":
                    return False
                steps = 0
                while abbr_ptr < len(abbr) and abbr[abbr_ptr].isdigit():
                    steps = (steps * 10) + int(abbr[abbr_ptr])
                    abbr_ptr += 1
                word_ptr += steps
            else:
                if abbr[abbr_ptr] != word[word_ptr]:
                    return False
                abbr_ptr += 1
                word_ptr += 1
        
        
        return word_ptr == len(word) and abbr_ptr == len(abbr)
            
        

File: 0408-valid-word-abbreviation/0408-valid-word-abbreviation.js
/**
 * @param {string} word
 * @param {string} abbr
 * @return {boolean}
 */
var validWordAbbreviation = function(word, abbr) {
    let wordPtr = 0;
    let abbrPtr = 0;
    
    while (wordPtr < word.length && abbrPtr < abbr.length) {
        if (!isNaN(abbr[abbrPtr])) {
            if (abbr[abbrPtr] === "0") {
                return false;
            }
            let steps = 0;
            while (abbrPtr < abbr.length && !isNaN(abbr[abbrPtr])) {
                steps = (steps * 10) + parseInt(abbr[abbrPtr]);
                abbrPtr++;
            }
            
            wordPtr += steps;
        } else {
            if (word[wordPtr] !== abbr[abbrPtr]) {
                return false;
            }
            wordPtr++;
            abbrPtr++;
        }
    }
    
    return wordPtr === word.length && abbrPtr === abbr.length; 
};